src/librustc_middle/ty/util.rs

```diff
@@ -779,6 +779,27 @@ impl<'tcx> ty::TyS<'tcx> {
         }
     }
 
+    /// Returns `true` if this type implements both `PartialStructuralEq` and `StructuralEq`.
+    ///
+    /// These marker traits are implemented along with `PartialEq` and `Eq` when implementations
+    /// for those traits are derived. They indicate that two values of this type are equal if all
+    /// of their fields are equal. A quirk of the marker traits is that their implementation is
+    /// never conditional on generic parameter bounds. For that reason, this helper function does
+    /// not take a `ParamEnv`.
+    ///
+    /// This function is "shallow" because it may return `true` for a type whose fields are not
+    /// `StructuralEq`. You will need to use a type visitor if you want to know whether a call to
+    /// `PartialEq::eq` will proceed structurally for a given type.
+    #[inline]
+    pub fn is_structural_eq_shallow(&'tcx self, tcx: TyCtxt<'tcx>) -> bool {
+        // Fast path for some builtin types
+        if self.is_primitive() || self.is_str() {
```

Comment:
This is incomplete at the moment, since many builtin types don't actually implement `StructuralEq` AFAIK but are handled specially in the type visitor. I don't know if we want to resolve this problem by checking for them here or by implementing the trait for various builtins (arrays, pointers, references, etc.)

Reply:
I suspect if we attempt to handle this by implementing the trait in all the cases of interest, we're going to run into the problem where we cannot implement it for types with a `for <'a> ...`, which arise from function pointers with a reference argument. (See also https://github.com/rust-lang/rust/issues/46989#issuecomment-353805481 ... and I'm struggling to find a specific issue about **this** particular wart in the language today...)

Reply:
Is it *bad* to leave them unhandled here for now? All that means is that you miss the fast path for those types, right? 

Then that is not a correctness concern; it is at most a performance concern, right? Which to me implies that we **can** merge this without figuring out the answer to your question.......

Reply:
So right now, this will return `false` for types such as `[i64; 4]` and `&str`. This is correct for now because `is_structural_eq_shallow` is only used in conjunction with type visitors that handles those cases separately and only invoke the query for `TyKind::Adt`. We could just document this fact and call it a day, or switch this query to accept only `TyKind::Adt`s, but I think it will be less prone to misuse if we handle the fundamental types inside the query instead of making the caller do it. 

There's two ways to accomplish this, we could either move this special-casing into the query itself or implement `StructuralEq` for fundamental types. Your point about function pointers suggests that we just do the first one, since we can't cover all the fundamental types via trait impls.

Reply:
@pnkfelix ^

Reply:
My main reaction is: I think you're rightfully worried about the method being prone to misuse.

 * In particular, I'm worried that the doc-comment you've put above `is_structurally_eq_shallow` might actually detract from readers' understanding of the code as written. (It adopt sa relatively simple specification ("when this returns true, you can conclude X; when it returns false, you cannot conclude anything"), which is what led me to say "can't we just land this code as is, and expand it later?"). But as you point out, all the current call-sites are making **stronger** assumptions about its behavior ("when given an ADT, then true implies X, and false implies not-X").

My first thought is: It would be nice to decide what interface we want, and then have the calls match it. But then my second thought is: It seems silly to put the effort into making all the call-sites handle the general case when they *know* that the method will produce usable results immediately when given an ADT.

In the long-term, you're probably right that we should move the special-casing into the query itself, so that it will handle any input `ty` properly.

However, I do not want to block this PR on that addition. For various reasons (such as ease of identifying source of performance regressions), I think we should land this PR as is, with an addition to the doc-comment saying that the `is_structurally_eq_*` methods happen to be precise when given an ADT. (i.e., I am advising that we indeed "just document this fact and call it a day", except I'm only saying "call it a day" for the short term; maybe add a FIXME for the longer term goal.)

Reply:
> (It adopts a relatively simple specification ("when this returns true, you can conclude X; when it returns false, you cannot conclude anything")

actually maybe you never intended this interpretation? I personally read into the first sentence quite literally as just "if" and **not** "if and only if".

In any case, I still recommend the approach of adding the comment explaining the situation, landing the PR, and leaving further generalization for later work.